### PR TITLE
Expose Stream Recording Apis in python

### PR DESCRIPTION
**Pull request by eellison**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #96775
* #89146
* __->__ #96384

Differential Revision: [D43999891](https://our.internmc.facebook.com/intern/diff/D43999891)

### CODE REVIEW

File: test/test_cuda.py [hunk 1]
```diff
@@ -5429,6 +5429,78 @@ def test_resnet(self):
...
+            all cudagraph tensors in use should be reflected in the allocator or they will be overwritten.
+            existing_graph should already have been used in a capture, and the mem_pool must already exist.
+            """
+            capture_id = existing_graph.id()
+            torch._C._cuda_allocateThreadToPrivatePool(device, mem_pool)
+            torch._C._cuda_notifyCaptureBegin(device, capture_id, mem_pool)
+            try:
+                yield
+            finally:
+                torch._C._cuda_notifyCaptureAboutToEnd(device, capture_id)
```
Comment (eellison) [inline test/test_cuda.py:5455]: We need a capture id unique identifier that wont conflict with any other ongoing unique identifiers. The way to generate a unique identifier is through starting a cudagraph stream recording in cuda. 

Here I am just reusing a finished recording id with the graph, we could alternately start and end a cudastream recording, no strong opinions here..

**Comment by davidberard98** [comment]
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D43999891).

File: torch/csrc/cuda/Module.cpp [hunk 2]
```diff
@@ -1203,6 +1203,25 @@ static void registerCudaPluggableAllocator(PyObject* module) {
     return c10::cuda::CUDACachingAllocator::getCheckpointState(device, id);
   });
 
+  m.def(
+      "_cuda_beginAllocateCurrentStreamToPool",
+      [](int device, at::cuda::MempoolId_t mempool_id) {
```
Comment (zdevito) [inline torch/csrc/cuda/Module.cpp:1208]: nit: feels weird to use the default stream but explicit device vs (default device + default stream) or (explicit stream)

File: test/test_cuda.py [hunk 3]
```diff
@@ -5511,6 +5511,81 @@ def test_resnet(self):
...
+
+
+        def no_pool():
+            # two allocations
+            for _ in range(10):
+                a = int8_cuda(LARGE_BUFFER)
+                b = int8_cuda(LARGE_BUFFER)
+                del a, b
+
+        graph_thread = threading.Thread(target=use_pool)
```
Comment (zdevito) [inline test/test_cuda.py:5570]: what is the intention of the threads here? In all likelihood without additional synchronization one is just going to run entirely before the other.  Barriers placed inside the threads would guarantee some interleaving.

**Review by zdevito — APPROVED** [review]
Looks good. Couple minor comments.

File: torch/csrc/cuda/Module.cpp (hunk 2 above)
Comment (eellison) [reply torch/csrc/cuda/Module.cpp:1049]: True. Is there any way of passing a python object as `CudaStream_t` ?

**Pull request closed without merging**